README.md
# iidy (Is it done yet?) -- a CloudFormation CLI tool

iidy improves the developer experience with CloudFormation and
CloudFormation templates.

* It provides immediate, readable feedback about what CloudFormation is doing
  and any errors it encounters.
* It is simple to learn, understand, and use. There are fewer moving parts and
  its commands map directly to CloudFormation's.
* It has simple, reliable support for AWS profiles.
* It supports the full range of CloudFormation operations, including
  _changesets_ without requiring any code beyond what is required to create a
  stack.
* It does some template validation, guards against common issues, and will be
  extended over time to validate our best practices and security policies.
* It provides seamless integration with AWS Parameter Store.
* It includes an optional YAML pre-processor which allows CloudFormation
  templates to be abstracted and simplified in ways not possible with vanilla
  CloudFormation templates. The pre-processor language supports importing data
  from a variety of external sources and this allows a better separation of
  concerns than is possible with stock CloudFormation without resorting to
  Lambda-backed custom resources. See the [pre-processor documentation
  below](#yaml-pre-processing) for more details.
* It has bash command completion support.

## Table of Contents

- [Pronunciation](#pronunciation)
- [Demo](#demo)
- [Installation](#installation)
    - [Binary Install for macOS via Homebrew](#binary-install-for-macos-via-homebrew)
    - [Binary Installation on Other Platforms](#binary-installation-on-other-platforms)
    - [Installation from Source](#installation-from-source)
    - [Docker Image](#docker-image)
- [Usage](#usage)
    - [Help](#help)
    - [The Args File](#the-args-file)
        - [Required Properties](#required-properties)
        - [Optional Properties](#optional-properties)
        - [Example](#example)
        - [Importing Data](#importing-data)
        - [Implicit Variables](#implicit-variables)
        - [Defining Variables](#defining-variables)
    - [AWS IAM Settings](#aws-iam-settings)
    - [Environment Variables](#environment-variables)
- [Examples](#examples)
- [Development](#development)
- [License](#license)
- [Releasing](#releasing)
- [Converting Existing CloudFormation Stacks to iidy](docs/converting-existing-cloudformation-stacks-to-iidy.md)
- [Custom Resource Templates](docs/custom-resource-templates.md)
- [Using Parameter Store](docs/using-parameter-store.md)
- [Working with Non-CloudFormation YAML](docs/working-with-non-cloudformation-yaml.md)
- [Working with StackSets](docs/working-with-stacksets.md)
- [YAML Processing](docs/yaml-preprocessing.md)
- [Extracting Data From Templates](docs/extracting-data-from-templates.md)

## Pronunciation

iidy is pronounced "eye-dee", like the audience's response to Cab Calloway in
[Minnie the Moocher](https://www.youtube.com/watch?v=8mq4UT4VnbE&feature=youtu.be&t=50s).

## Demo

[![asciicast](https://asciinema.org/a/8rzW1WyoDxMdVJpvpYf2mHm8E.png)](https://asciinema.org/a/8rzW1WyoDxMdVJpvpYf2mHm8E)

## Installation

iidy is distributed as a self-contained executable (via [pkg](https://github.com/zeit/pkg)).

### Binary Install for macOS via Homebrew

Use Unbounce's custom Homebrew Tap to install iidy. This is the preferred method for macOS.

```shell
brew tap unbounce/homebrew-taps
brew update
brew install iidy
```

### Binary Installation on Other Platforms

```shell
# Grab the appropriate binary from the releases page.
wget https://github.com/unbounce/iidy/releases/download/v1.6.7/iidy-linux-amd64.zip
# or wget https://github.com/unbounce/iidy/releases/download/v1.6.7/iidy-macos-amd64.zip
unzip iidy*.zip
chmod +x iidy
mv iidy /usr/local/bin/ # or somewhere more appropriate
```

### Installation from Source

Counter-intuitively, this requires more disk space than the binary
install. You need Node 7+ and npm 4+ installed.

```shell
git clone git@github.com:unbounce/iidy.git
cd iidy
npm install && npm run build # to compile our source first
ln -s $(pwd)/bin/iidy /usr/local/bin/
# or npm install -g .
```

## Usage

### Help

```
$ iidy help
iidy - CloudFormation with Confidence                    An acronym for "Is it done yet?"

Commands:
  iidy create-stack     <argsfile>                            create a cfn stack based on stack-args.yaml
  iidy update-stack     <argsfile>                            update a cfn stack based on stack-args.yaml
  iidy create-or-update <argsfile>                            create or update a cfn stack based on stack-args.yaml
  iidy estimate-cost    <argsfile>                            estimate aws costs based on stack-args.yaml

  iidy create-changeset           <argsfile> [changesetName]  create a cfn changeset based on stack-args.yaml
  iidy exec-changeset             <argsfile> <changesetName>  execute a cfn changeset based on stack-args.yaml

  iidy describe-stack      <stackname>                        describe a stack
  iidy watch-stack         <stackname>                        watch a stack that is already being created or updated
  iidy describe-stack-drift <stackname>                       describe stack drift
  iidy delete-stack        <stackname>                        delete a stack (after confirmation)
  iidy get-stack-template  <stackname>                        download the template of a live stack
  iidy get-stack-instances <stackname>                        list the ec2 instances of a live stack
  iidy list-stacks                                            list all stacks within a region

  iidy param                                                  sub commands for working with AWS SSM Parameter Store

  iidy template-approval                                      sub commands for template approval

  iidy render <template>                                      pre-process and render yaml template
  iidy get-import <import>                                    retrieve and print an $import value directly
  iidy demo   <demoscript>                                    run a demo script
  iidy lint-template   <argsfile>                             lint a CloudFormation template
  iidy convert-stack-to-iidy <stackname> <outputDir>          create an iidy project directory from an existing CFN stack
  iidy init-stack-args                                        initialize stack-args.yaml and cfn-template.yaml

  iidy completion                                             generate bash completion script. To use: "source <(iidy completion)"

AWS Options:
  --client-request-token  a unique, case-sensitive string of up to 64 ASCII characters used to ensure idempotent retries.         [string] [default: null]
  --region                AWS region. Can also be set via --environment & stack-args.yaml:Region.                                 [string] [default: null]
  --profile               AWS profile. Can also be set via --environment & stack-args.yaml:Profile. Use --profile=no-profile to override values in
                          stack-args.yaml and use AWS_* env vars.                                                                 [string] [default: null]
  --assume-role-arn       AWS role. Can also be set via --environment & stack-args.yaml:AssumeRoleArn. This is mutually exclusive with --profile. Use
                          --assume-role-arn=no-role to override values in stack-args.yaml and use AWS_* env vars.                 [string] [default: null]

Options:
  --environment, -e  used to load environment based settings: AWS Profile, Region, etc.                                  [string] [default: "development"]
  --debug            log debug information to stderr.                                                                           [boolean] [default: false]
  --log-full-error   log full error information to stderr.                                                                      [boolean] [default: false]
  -v, --version      show version information                                                                                                    [boolean]
  -h, --help         show help                                                                                                                   [boolean]

Status Codes:
  Success (0)       Command successfully completed
  Error (1)         An error was encountered while executing command
  Cancelled (130)   User responded 'No' to iidy prompt or interrupt (CTRL-C) was received
```

### The Args File

Metadata for a CloudFormation stack is kept in a file called the "argsfile"
(typically called `stack-args.yaml`). This file is a parameter for commands like
`iidy create-or-update` or `iidy create-changeset` and is the main source of
data for creating or updating a CloudFormation stack.

#### Required Properties

| Property | Description | Example |
|----------|-------------|---------|
| StackName | CloudFormation stack name | `my-stack` |
| Template | Local, `https` or `s3` location of CloudFormation template | `cfn-template.yaml` |

#### Optional Properties

| Property | Description | Example |
|----------|-------------|---------|
| Region | AWS Region to use, overridden by `--region` and `AWS_REGION` | `us-east-1` |
| Profile | [AWS profile](https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-profiles.html) to use, overridden by `--profile` and `AWS_PROFILE` |
| AssumeRoleARN | IAM Role ARN to assume | `arn:aws:iam::1234567890:role/my-role` |
| ServiceRoleARN | [CloudFormation Service Role](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/using-iam-servicerole.html) to use |
| Tags | Tags to assign to CloudFormation stack | `{ service: my-app }` |
| Parameters | [CloudFormation stack parameters](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/parameters-section-structure.html) to use |
| Capabilities | List of [CloudFormation capabilities to use](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/using-iam-template.html#using-iam-capabilities) | `CAPABILITY_NAMED_IAM` |
| NotificationARNs | List of [SNS Topic ARNs to send CloudFormation notifications to] | `arn:aws:sns:us-east-1:123467890:my-topic` |
| TimeoutInMinutes | Number of minutes to provide as [timeout to the CloudFormation service](https://docs.aws.amazon.com/AWSCloudFormation/latest/APIReference/API_CreateStack.html) | `10` |
| OnFailure | options: `ROLLBACK`, `DELETE`, `DO_NOTHING`, default `ROLLBACK` | `DELETE`
| DisableRollback | [boolean defaulting to false](https://docs.aws.amazon.com/AWSCloudFormation/latest/APIReference/API_CreateStack.html)

CHANGELOG.md
### Changelog

#### [v1.12.0](https://github.com/unbounce/iidy/compare/v1.11.0...v1.12.0)

> 29 September 2021

- PQ-229 Use github actions, report code coverage to sonar [`#265`](https://github.com/unbounce/iidy/pull/265)
- PQ-228 Update sonar docker image and use properties file [`#264`](https://github.com/unbounce/iidy/pull/264)
- Add !$join function [`#249`](https://github.com/unbounce/iidy/pull/249)
- Add handlebars-helpers string helpers [`#251`](https://github.com/unbounce/iidy/pull/251)
- Touches up some spelling and grammar issues [`#244`](https://github.com/unbounce/iidy/pull/244)
- Avoid calling fs.statSync when input is from stdin [`#260`](https://github.com/unbounce/iidy/issues/260)
- Fixes spelling of `command-implementations` [`fe8386a`](https://github.com/unbounce/iidy/commit/fe8386a6ed6d6324aa596a6f0b4fa487bb04c3bf)
- Touches up some spelling/grammar in documentation [`fb52564`](https://github.com/unbounce/iidy/commit/fb52564747a14c736f8ea0fb078662000291bb19)
- fix spelling mistake in the README [`567a66e`](https://github.com/unbounce/iidy/commit/567a66e7f4a5fe612d4d8ad45d4bcaa9b7bbdaa8)
- add DELETE_SKIPPED as a colorized status [`d62b766`](https://github.com/unbounce/iidy/commit/d62b766881a207cbaae5a1eb18483d9c227e855b)
- Accept derived values for !$split and !$join [`39ed8fd`](https://github.com/unbounce/iidy/commit/39ed8fd4732e201fe23bf32522aa3a18873d746a)
- Add documentation [`a8fb86c`](https://github.com/unbounce/iidy/commit/a8fb86c7c116d6f4421f76c8a5a90e1d1045b539)
- Add toJson, toYaml alias, accept non literal values [`5ec7d6f`](https://github.com/unbounce/iidy/commit/5ec7d6fb373d370298185937dcd69558e0bff6c3)
- Bump lodash from 4.17.15 to 4.17.19 [`83ef7f6`](https://github.com/unbounce/iidy/commit/83ef7f6dabff68838bfa08fe72d5a34870eb93be)
- Use toJsonString/toYamlString [`1b9299a`](https://github.com/unbounce/iidy/commit/1b9299ab960575610a98408de95f382e3b0a4cc6)
- Clarify what is happening with trailing characters [`afa9de3`](https://github.com/unbounce/iidy/commit/afa9de3d0ee0d70823e62504b560d4a792cd34cf)
- Add note about purpose of !$split and !$join [`4acdfa0`](https://github.com/unbounce/iidy/commit/4acdfa092fcc9e529a1131352c7ca032489a1f9d)
- Deprecate non-camelCase helpers [`863b0e9`](https://github.com/unbounce/iidy/commit/863b0e9c034acc0cdfff3a7de917a743fd3116a1)
- Use S3 template object limit when doing template approval request [`8926290`](https://github.com/unbounce/iidy/commit/892629099a4d719cf74d57809f95831a5d6ccbc9)
- Bump node-fetch from 2.6.0 to 2.6.1 [`8d7b662`](https://github.com/unbounce/iidy/commit/8d7b6623ac426b1c76c8206c6d455177cc28f1a5)
- Bump dot-prop from 4.2.0 to 4.2.1 [`bd9b402`](https://github.com/unbounce/iidy/commit/bd9b402716c3fab12e1e4ed2171ecbf4a84d5cb5)
- bump laundry to v1.6.0 [`944aa66`](https://github.com/unbounce/iidy/commit/944aa66f6c9e93b16e7f6aac64fbc3f8129fe910)
- updates s3 template size to match new max of 1MB; updates using s3 template size when loading a template for approval since the template will be stored on s3 [`db225b8`](https://github.com/unbounce/iidy/commit/db225b8c09ddb06b38d27b741ab68ba670993dc8)
- bump deps and update a few exception types as required for ts update [`abd19a9`](https://github.com/unbounce/iidy/commit/abd19a9fa3cd93ba0f101d44f9e5a6f745430e89)
- update aws-sdk-js version (with our creds cache patch) [`a6b375d`](https://github.com/unbounce/iidy/commit/a6b375dec6cd6c6cd76478801e588544b938aa84)
- bump node version to 14x [`63747d4`](https://github.com/unbounce/iidy/commit/63747d4a5c0480a4f2507ff05ae403a3710a8719)
- update gen-changelog script and deps [`c7ec9aa`](https://github.com/unbounce/iidy/commit/c7ec9aa0a886ddcc5a6ffa4c2d9e96256a956b44)

#### [v1.11.0](https://github.com/unbounce/iidy/compare/v1.10.0...v1.11.0)

> 9 January 2020

- PR #235 by @tavisrudd - implement 145: display resource creation times
- PR #243 by @tavisrudd - fix issue #238: partial writes to stdout when piping render output
- PR #237 by @tavisrudd - Create CODE_OF_CONDUCT.md
- PR #240 by @dannosaur - #239 fix invalid refs when node objects are duplicated within the template
- PR #242 by @jpb - Support cross account template approval requests
- PR #232 by @tavisrudd - update build deps, improve `make prepare_release`, implement new changelog generator
- PR #233 by @tavisrudd - fix 228 and 105 double processing of refs in custom template Props
- PR #234 by @tavisrudd - remove references to yarn
- PR #231 by @kalupa - Add support for .nvmrc for node versioning
- PR #230 by @tavisrudd - disable autochangelog for now
- PR #229 by @tavisrudd - bump laundry to v1.5.0
- PR #227 by @tavisrudd - fix #178: shasum filehash portability
- PR #226 by @tavisrudd - add test for tojsonPretty handlebars function
- PR #221 by @tavisrudd - misc error message improvements
- PR #225 by @tavisrudd - fix #213: bug in resource node preprocessing
- PR #224 by @dannosaur - [#223] fix issue with $merge syntax in Resources
- PR #222 by @tavisrudd - add --color cli option

#### [v1.10.0](https://github.com/unbounce/iidy/compare/v1.9.0...v1.10.0)

> 6 December 2019

- Add parameter store global configuration to disable template approval [`#218`](https://github.com/unbounce/iidy/pull/218)

src/main.ts
import * as _ from 'lodash';
import * as path from 'path';
import * as fs from 'fs';
import * as process from 'process';

// reuse http connections to aws to improve perf. Set it here before import of aws.
process.env.AWS_NODEJS_CONNECTION_REUSE_ENABLED = '1';

const awsUserDir = process.env.HOME ? path.join(process.env.HOME as string, '.aws') : null;
if (awsUserDir && fs.existsSync(awsUserDir)) {
  // We need to set this early because of https://github.com/aws/aws-sdk-js/pull/1391
  // We also set this env-var in the main cli entry-point
  process.env.AWS_SDK_LOAD_CONFIG = '1'; // see https://github.com/aws/aws-sdk-js/pull/1391
  // Note:
  // if this is set and ~/.aws doesn't exist we run into issue #17 as soon as the sdk is loaded:
  //  Error: ENOENT: no such file or directory, open '.../.aws/credentials
}

// Use bluebird promises globally. We need to load this prior to 'aws-sdk'
import * as bluebird from 'bluebird';
global.Promise = bluebird;

import * as yargs from 'yargs';
import * as cli from 'cli-color';
import {Handler,
        description,
        fakeCommandSeparator,
        wrapCommandHandler,
        stackNameOpt,
        lintTemplateOpt} from './cli/utils';
import {Commands} from './cli/command-types';

// TODO bring these two in line with the new lazy load scheme
import {buildParamCommands} from './params/cli'
import {buildApprovalCommands} from './cfn/approval/cli'

// We lazy load the actual command fns to make bash command completion
// faster. See the git history of this file to see the non-lazy form.
// Investigate this again if we can use babel/webpack to shrinkwrap

export function lazyLoad(fnname: keyof Commands): Handler {
  return (args) => {
    const {implementations} = require('./cli/command-implementations');
    return implementations[fnname](args);
  }
}

function lazyGetter(target: Commands, key: keyof Commands) {
  Object.defineProperty(target, key, {value: lazyLoad(key)});
}

class LazyCommands implements Commands {
  @lazyGetter createStackMain: Handler
  @lazyGetter createOrUpdateStackMain: Handler
  @lazyGetter updateStackMain: Handler
  @lazyGetter listStacksMain: Handler
  @lazyGetter watchStackMain: Handler
  @lazyGetter describeStackMain: Handler
  @lazyGetter describeStackDriftMain: Handler
  @lazyGetter getStackTemplateMain: Handler
  @lazyGetter getStackInstancesMain: Handler
  @lazyGetter deleteStackMain: Handler
  @lazyGetter createChangesetMain: Handler
  @lazyGetter executeChangesetMain: Handler
  @lazyGetter lintMain: Handler
  @lazyGetter renderMain: Handler
  @lazyGetter getImportMain: Handler
  @lazyGetter estimateCost: Handler
  @lazyGetter demoMain: Handler
  @lazyGetter convertStackToIIDY: Handler
  @lazyGetter initStackArgs: Handler
}

const environmentOpt: yargs.Options = {
  type: 'string', default: 'development',
  alias: 'e',
  description: description('used to load environment based settings: AWS Profile, Region, etc.')
};

export function buildArgs(commands = new LazyCommands(), wrapMainHandler = wrapCommandHandler) {
  const usage = (`${cli.bold(cli.green('iidy'))} - ${cli.green('CloudFormation with Confidence')}`
    + ` ${' '.repeat(18)} ${cli.blackBright('An acronym for "Is it done yet?"')}`);
  const epilogue = ('Status Codes:\n'
    + '  Success (0)       Command successfully completed\n'
    + '  Error (1)         An error was encountered while executing command\n'
    + '  Cancelled (130)   User responded \'No\' to iidy prompt or interrupt (CTRL-C) was received');

  return yargs
    .scriptName('iidy')
    .env('IIDY')
    .command(
      'create-stack     <argsfile>',
      description('create a cfn stack based on stack-args.yaml'),
      (args) => args
        .demandCommand(0, 0)
        .usage('Usage: iidy create-stack <stack-args.yaml>')
        .option('stack-name', stackNameOpt)
        .option('lint-template', lintTemplateOpt(false)),
      wrapMainHandler(commands.createStackMain))

    .command(
      'update-stack     <argsfile>',
      description('update a cfn stack based on stack-args.yaml'),
      (args) => args
        .demandCommand(0, 0)
        .usage('Usage: iidy update-stack <stack-args.yaml>')
        .option('stack-name', stackNameOpt)
        .option('lint-template', lintTemplateOpt(false))
        .option('changeset', {
          type: 'boolean', default: false,
          description: description('review & confirm changes via a changeset')
        })
        .option('yes', {
          type: 'boolean', default: false,
          description: description('Confirm and execute changeset if --changeset option is used')
        })
        .option('diff', {
          type: 'boolean', default: true,
          description: description('diff & review changes to the template body as part of changeset review')
        })
        .option('stack-policy-during-update', {
          type: 'string', default: null,
          description: description('override original stack-policy for this update only')
        }),
      wrapMainHandler(commands.updateStackMain))

    .command(
      'create-or-update <argsfile>',
      description('create or update a cfn stack based on stack-args.yaml'),
      (args) => args
        .demandCommand(0, 0)
        .usage('Usage: iidy create-or-update <stack-args.yaml>')
        .option('stack-name', stackNameOpt)
        .option('lint-template', lintTemplateOpt(false))
        .option('changeset', {
          type: 'boolean', default: false,
          description: description('review & confirm changes via a changeset')
        })
        .option('yes', {
          type: 'boolean', default: false,
          description: description('Confirm and execute changeset if --changeset option is used')
        })
        .option('diff', {
          type: 'boolean', default: true,
          description: description('diff & review changes to the template body as part of changeset review')
        })
        .option('stack-policy-during-update', {
          type: 'string', default: null,
          description: description('override original stack-policy for this update only')
        }),
      wrapMainHandler(commands.createOrUpdateStackMain))

    .command(
      'estimate-cost    <argsfile>',
      description('estimate aws costs based on stack-args.yaml'),
      (args) => args
        .demandCommand(0, 0)
        .option('stack-name', stackNameOpt),
      wrapMainHandler(commands.estimateCost))

    .command(fakeCommandSeparator, '')

    .command(
      'create-changeset           <argsfile> [changesetName]',
      description('create a cfn changeset based on stack-args.yaml'),
      (args) => args
        .demandCommand(0, 0)
        .option('watch', {
          type: 'boolean', default: false,
          description: description('Watch stack after creating changeset. This is useful when exec-changeset is called by others.')
        })
        .option('watch-inactivity-timeout', {
          type: 'number', default: (60 * 3),
          description: description('how long to wait for events when the stack is in a terminal state')
        })
        .option('description', {
          type: 'string', default: undefined,
          description: description('optional description of changeset')
        })
        .option('stack-name', stackNameOpt),
      wrapMainHandler(commands.createChangesetMain))

    .command(
      'exec-changeset             <argsfile> <changesetName>',
      description('execute a cfn changeset based on stack-args.yaml'),
      (args) => args
        .demandCommand(0, 0)
        .option('stack-name', stackNameOpt),
      wrapMainHandler(commands.executeChangesetMain))

    .command(fakeCommandSeparator, '')

    .command(
      'describe-stack      <stackname>',
      description('describe a stack'),
      (args) => args
        .demandCommand(0, 0)
        .option('events', {
          type: 'number', default: 50,
          description: description('how many stack events to display')
        })
        .option('query', {
          type: 'string', default: null,
          description: description('jmespath search query to select a subset of the output')
        })
        .usage('Usage: iidy describe-stack <stackname-or-argsfile>'),
      wrapMainHandler(commands.describeStackMain))

    .command(
      'watch-stack         <stackname>',
      description('watch a stack that is already being created or updated'),
      (args) => args
        .demandCommand(0, 0)
        .option('inactivity-timeout', {
          type: 'number', default: (60 * 3),
          description: description('how long to wait for events when the stack is in a terminal state')
        })
        .usage('Usage: iidy watch-stack <stackname-or-argsfile>'),
      wrapMainHandler(commands.watchStackMain))

    .command(
      'describe-stack-drift <stackname>',
      description('describe stack drift'),
      (args) => args
        .demandCommand(0, 0)
        .option('drift-cache', {
          type: 'number', default: (60 * 5),
          description: description('how long to cache previous drift detection results (seconds)')
        })
        .usage('Usage: iidy describe-stack-drift <stackname-or-argsfile>'),
      wrapMainHandler(commands.describeStackDriftMain))

    .command(
        'delete-stack        <stackname>',
        description('delete a stack (after confirmation)'),
        (args) => args
          .demandCommand(0, 0)
          .option('role-arn', {
            type: 'string',
            description: description('Role to assume for delete operation')
          })
          .option('retain-resources', {
            type: 'string',
            array: true,
            description: description('For stacks in the DELETE_FAILED, list of logical resource ids to retain')
          })
          .option('yes', {
            type: 'boolean', default: false,
            description: description('Confirm deletion of stack')
          })
          .option('fail-if-absent', {
            type: 'boolean', default: false,
            description: description('Fail if stack is absent (exit code = 1). Default is to tolerate absence.')
          })
          .usage('Usage: iidy delete-stack <stackname-or-argsfile>'),
        wrapMainHandler(commands.deleteStackMain))

        .command(
          'get-stack-template  <stackname>',
          description('download the template of a live stack'),
          (args) => args
            .demandCommand(0, 0)
            .option('format', {
              type: 'string', default: 'original',
              choices: ['original', 'yaml', 'json'],
              description: description('Template stage to show')
            })
            .option('stage', {
              type: 'string', default: 'Original',
              choices: ['Original', 'Processed'],
              description: description('Template stage to show')
            })
            .usage('Usage: iidy get-stack-template <stackname-or-argsfile>'),
          wrapMainHandler(commands.getStackTemplateMain))

        .command(
          'get-stack-instances <stackname>',
          description('list the ec2 instances of a live stack'),
          (args) => args
            .demandCommand(0, 0)
            .option('short', {
              type: 'boolean', default: false,
              description: description('Show only instance dns names')
            })
            .usage('Usage: iidy get-stack-instances <stackname-or-argsfile>'),
          wrapMainHandler(commands.getStackInstancesMain))

        .command(
          'list-stacks',
          description('list all stacks within a region'),
          (args) => args
            .demandCommand(0, 0)
            .option('tag-filter', {
              type: 'string', default: [],
              array: true,
              description: description('Filter by tags: key=value')
            })
            .option('jmespath-filter', {
              type: 'string', default: null,
              description: description('jmespath search query to select a subset of the stacks')
            })
            .option('query', {
              type: 'string', default: null,
              description: description('jmespath search query to select a subset of the output')
            })
            .option('tags', {
              type: 'boolean', default: false,
              description: description('Show stack tags')
            }),
          wrapMainHandler(commands.listStacksMain))

        .command(fakeCommandSeparator, '')

        .command('param',
          description('sub commands for working with AWS SSM Parameter Store'),
          buildParamCommands)

        .command(fakeCommandSeparator, '')

        .command('template-approval',
          description('sub commands for template approval'),
          buildApprovalCommands)

        .command(fakeCommandSeparator, '')

        .command(
          'render <template>',
          description('pre-process and render yaml template'),
          (args) => args
            .demandCommand(0, 0)
            .usage('Usage: iidy render <input-template.yaml>')
            .positional('template', {
              description: 'template file to render, `-` for STDIN, can also be a directory of templates (will only render *.yml and *.yaml files in directory)'
            })
            .option('outfile', {
              type: 'string', default: 'stdout',
              description: description('yaml input template to preprocess')
            })
            .option('format', {
              type: 'string', default: 'yaml',
              choices: ['yaml', 'json'],
              description: description('output serialization syntax')
            })
            .option('query', {
              type: 'string', default: null,
              description: description('jmespath search query to select a subset of the output')
            })
            .option('overwrite', {
              type: 'boolean', default: false,
              description: description('Whether to overwrite an existing <outfile>.')
            })
            .strict(),
          wrapMainHandler(commands.renderMain))

        .command(
          'get-import <import>',
          description('retrieve and print an $import value directly'),
          (args) => args
            .demandCommand(0, 0)
            .usage('Usage: iidy get-import <import>')
            .option('format', {
              type: 'string', default: 'yaml',
              choices: ['yaml', 'json'],
              description: description('output serialization syntax')
            })
            .option('query', {
              type: 'string', default: null,
              description: description('jmespath search query to select a subset of the output')
            })
            .strict(),
          wrapMainHandler(commands.getImportMain))

        .command(
          'demo   <demoscript>',
          description('run a demo script'),
          (args) => args
            .demandCommand(0, 0)
            .option('timescaling', {
              type: 'number', default: 1,
              description: description('time scaling factor for sleeps, etc.')
            })
            .strict(),
          wrapMainHandler(commands.demoMain))
        .command(
          'lint-template   <argsfile>',
          description('lint a CloudFormation template'),
          (args) => args
            .demandCommand(0, 0)
            .option('use-parameters', {
              type: 'boolean', default: false,
              description: description('use parameters to improve linting accuracy')
            })
            .strict(),
          wrapMainHandler(commands.lintMain))
        .command(
          'convert-stack-to-iidy <stackname> <outputDir>',
          description('create an iidy project directory from an existing CFN stack'),
          (args) => args
            .demandCommand(0, 0)
            .option('move-params-to-ssm', {
              type: 'boolean', default: false,
              description: description('automatically create an AWS SSM parameter namespace populated with the stack Parameters')
            })
            .option('sortkeys', {
              type: 'boolean', default: true,
              description: description('sort keys in cloudformation maps')
            })
            .option('project', {
              type: 'string', default: null,
              description: description('The name of the project (service or app). If not specified the "project" Tag is checked.')
            }),
          wrapMainHandler(commands.convertStackToIIDY))

        .command(
          'init-stack-args',
          description('initialize stack-args.yaml and cfn-template.yaml'),
          (args) => args
            .demandCommand(0, 0)
            .option('force', {
              type: 'boolean', default: false,
              description: description('Overwrite the current stack-args.yaml and cfn-template.yaml')
            })
            .option('force-stack-args', {
              type: 'boolean', default: false,
              description: description('Overwrite the current stack-args.yaml')
            })
            .option('force-cfn-template', {
              type: 'boolean', default: false,
              description: description('Overwrite the current cfn-template.yaml')
            }),
          wrapMainHandler(commands.initStackArgs))

        .option('environment', environmentOpt)
        .option('client-request-token', {
          type: 'string', default: null,
          group: 'AWS Options:',
          description: description('a unique, case-sensitive string of up to 64 ASCII characters used to ensure idempotent retries.')
        })
        .option('region', {
          type: 'string', default: null,
          group: 'AWS Options:',
          description: description('AWS region. Can also be set via --environment & stack-args.yaml:Region.')
        })
        .option('profile', {
          type: 'string', default: null,
          group: 'AWS Options:',
          description: description(
            'AWS profile. Can also be set via --environment & stack-args.yaml:Profile. '
            + 'Use --profile=no-profile to override values in stack-args.yaml and use AWS_* env vars.')
        })
        .option('assume-role-arn', {
          type: 'string', default: null,
          group: 'AWS Options:',
          description: description(
            'AWS role. Can also be set via --environment & stack-args.yaml:AssumeRoleArn. '
            + 'This is mutually exclusive with --profile. '
            + 'Use --assume-role-arn=no-role to override values in stack-args.yaml and use AWS_* env vars.')
        })

        .option('color', {
          type: 'string', default: 'auto',
          choices: ['auto', 'always', 'never'],
          description: description('whether to color output using ANSI escape codes')
        })
        .middleware((args) => {
          if (args.color === 'never' || (args.color === 'auto' && ! process.stdout.columns)) {
            process.env.NO_COLOR='1';
          }
        })
        .option('debug', {
          type: 'boolean', default: false,
          description: description('log debug information to stderr.')
        })
        .option('log-full-error', {
          type: 'boolean', default: false,
          description: description('log full error information to stderr.')
        })
        .command(fakeCommandSeparator, '')

        .demandCommand(1)
        .usage(usage)
        .version()
        .alias('v', 'version')
        .describe('version', description('show version information'))
        .help('help', description('show help'))
        .alias('h', 'help')
        .completion('completion', description('generate bash completion script. To use: "source <(iidy completion)"'))
        .recommendCommands()
        .epilogue(epilogue)
        .strict()
        .wrap(yargs.terminalWidth());
}

export async function main() {
  // called for side-effect to force parsing / handling
  // tslint:disable-next-line
  buildArgs().argv; // NOSONAR
}


if (module.parent === null) {
  main();
};
